File: android/app/src/main/kotlin/com/iblocker/android/upstream/UpstreamFactory.kt
package com.iblocker.android.upstream

import com.iblocker.core.engine.DnsUpstream
import com.iblocker.core.shared.UpstreamConfig

/** Builds the concrete upstream for a stored configuration. */
object UpstreamFactory {
    fun make(config: UpstreamConfig, protector: SocketProtector = SocketProtector.NONE): DnsUpstream? =
        when (config.kind) {
            UpstreamConfig.Kind.DOH -> config.dohURL
                ?.takeIf { it.startsWith("https://") }
                ?.let { DohUpstream(it, protector) }

            UpstreamConfig.Kind.UDP -> config.udpAddress
                ?.takeIf { it.isNotBlank() }
                ?.let { UdpUpstream(it, protector = protector) }
        }
}

File: ios/IBlockerKit/Sources/IBlockerTunnelKit/UpstreamFactory.swift
import Foundation
import IBlockerKit

/// Builds the concrete upstream for a stored configuration.
public enum UpstreamFactory {
    public static func make(_ config: UpstreamConfig) -> DNSUpstream? {
        switch config.kind {
        case .doh:
            return DoHUpstream(config: config)
        case .udp:
            #if canImport(Network)
            guard let address = config.udpAddress, !address.isEmpty else { return nil }
            return UDPUpstream(address: address)
            #else
            return nil
            #endif
        }
    }
}
